rfd/0177-upgrade-js-package-manager.md

```diff
@@ -0,0 +1,185 @@
+still type `yarn start-teleport`. 
+However, if we are going to use pnpm installation method, why not switch to pnpm 
+entirely? 
+Yarn also shows a warning `The pnpm linker doesn't support providing different 
+versions to workspaces' peer dependencies` during the installation, but I'm not 
+sure what significance this has.
+
+To sum up: I recommend migrating to pnpm.
+
+## Migration process
+### Establishing correct relationships between workspaces
+After switching the package manager to pnpm, the Web UI/Connect apps won’t run, 
+due to TS/JS code in the workspaces not being able to resolve their imports 
+(like `react`).
+Why? This is a continuation of the hoisting problem — our workspaces rely 
+on dependencies of other workspaces, without declaring them in their 
+`package.json`.
+It worked in Yarn 1.x, since adding a dependency to a one workspace made it 
+available in any other workspace (because of hoisting to the root 
+`node_modules`).
+We can see this in `packages/build`: the `@opentelemetry/*` dependencies are 
+declared there but used only in `packages/teleport`.
+
+This contradicts the idea of workspaces: they are designed to encapsulate 
+project dependencies and code. Sharing dependencies between unrelated workspaces 
+breaks this encapsulation, leading to potential conflicts and contamination 
+where changes in one workspace inadvertently affect another.
+
+To allow the projects to compile again, the dependencies need to be moved to the
+correct workspaces.
+I see three scenarios for how they should be moved over:
+1. Dependencies that are used only in a single workspace should be kept there.
+Example: `electron-builder` that builds Connect - it’s declared in 
+`packages/build`, but it should be in `packages/teleterm`.
+2. Project-wide dependencies imported everywhere, should be kept in the root 
+`package.json`.
+Example: `react` - declared in `build/package.json`, but imported in every other 
+workspace.
+3. Project-wide tools used in the root `package.json` should be kept in 
+`build/package.json`, if possible. 
+Example: `jest` or `storybook` can be installed in `build/package.json` 
+which will expose them through scripts to the root `package.json`.
+
+There is also an edge case for teleport.e. Its dependencies are declared in 
+`packages/e-imports`, allowing them to be installed when only the OSS teleport 
+repository is cloned. This makes the lockfile always look the same, regardless 
+of teleport.e being cloned or not. 
+After our workspaces become “isolated”, code in teleport.e will no longer be 
+able to import these dependencies. 
+This problem can be solved by re-exporting them from `packages/e-imports`.
+For example, `import Highlight from ‘react-highlight’` can become 
+`import Highlight from '@gravitational/e-imports/react-highlight'`.
+The ergonomics of this solution are not ideal, but the only other option is to 
+move these dependencies to the root `package.json`.
```

Comment:
Seems simpler to keep it in the root `package.json`

It's only a handful of libraries and it means we don't have a confusing difference between should it be imported `@gravitational/e-imports/` or not, for the sake of keeping some dependencies in a different JSON file

Reply:
I think you're right, I wanted to avoid polluting root `package.json` with unneeded dependencies, but re-exporting the packages from `e-imports` may be too cumbersome.
And as you said, there shouldn't be too many libraries used exclusively by teleport.e.

I will update the RFD.

Reply:
I don't like the idea of putting package-specific deps in the root, but I also don't see a good way of addressing all of our concerns at the same time. 🤔

If we put deps for a specific package in the root package.json, what stops us from just putting all of them there? It'd be easy for some to put teleport or teleterm specific packages in there as well. From their perspective, all they would see is that they had followed an established pattern. There's no easy way to detect that and we'd need to enforce it manually.

Using e-imports would actually enforce that all enterprise-only deps are added to the OSS first and then consumed in ent. It also eliminates the possibility of things getting out of sync. But exposing deps as `@gravitational/e-imports/foo` certainly adds a bunch of work whenever someone wants to add a dep. How would we even do that? Require someone to set up `foo/index.ts` in `e-imports`? idk if there's an easier way.

We could also duplicate ent-only deps in both teleport and teleport.e. That would make them easier to consume. But then we make it easy to update only one of the package.jsons.

Do you have any other ideas?

Reply:
>How would we even do that? Require someone to set up `foo/index.ts` in e-imports? idk if there's an easier way.

Yeah, something like that.

I agree with everything you wrote here. I'm not super happy with any of these options, unfortunately I couldn't find anything better. 
But I will spend some more time tomorrow researching it, maybe I will find something!

Reply:
Actually, there is one more thing that comes to my mind: having a separate `pnpm-lock.yaml` in each workspace. 
The config option is called [`shared-workspace-lockfile`](https://pnpm.io/npmrc#shared-workspace-lockfile). 
I think it would provide the best DX when it comes to `e` dependencies - if you need a new package, you don't have to touch OSS at all. We would get rid of `e-imports` completely. 
Also, it's easier to review changes in a smaller lockfile, rather than in a one having 30K+ lines.

But there are some disadvantages too:
- slower installation
- dependabot, we would have to add an entry for JS deps in `e` (and I'm not sure if OSS would work with no changes)

It seems to me that it would give us the best of both worlds (strictness and good developer experience), but OTOH I'm a bit afraid of making our setup more complex.

Reply:
> slower installation

How much slower? I think we can afford a bit more time for the sake of developer experience

I think this is a good option and we should try it out, as long as it doesn't slow us down too much. If it isn't great, we can try other methods

Reply:
Here are some quick tests. The difference in noticeable but not huge:
single lockfile: 8 s 
separate lockfiles: 13.7 s

I ran `make clean-ui` first and then `pnpm i`. 
I didn't clear the global store, so dependencies were resolved locally without downloading them from npm.

For comparison, in the same test, it took around 15 s for Yarn 1.x to install the deps, so we would be comparably fast.

Reply:
I had an inkling that this is not how the majority of people use pnpm, especially those that work on pnpm itself. I went to pnpm's repo and there's [quite a few issues with `shared-workspace-lockfile` in the title](https://github.com/pnpm/pnpm/issues?page=1&q=is%3Aissue+is%3Aopen+shared-workspace-lockfile). That's the biggest disadvantage IMHO. We'd be pretty much to square one with how workspaces were treated as experimental in Yarn v1.

Reply:
Out of all those options, I feel like putting teleport.e deps into the root of package.json might be the best option for now. While not the prettiest and having the potential to corrode the codebase over time, it introduces the least amount of friction and lets us move forward.

Down the road (I hope) we will invert the relationship between OSS and teleport.e. This has been brought up a couple of times and the way I understand it is that instead of teleport.e being a submodule of OSS, OSS would be a submodule of teleport.e.

This would solve the problem with needing a separate package for ent deps. OTOH, we will then have to solve the problem of having two separate lockfiles for each repo.

Reply:
That's a good point about the `shared-workspace-lockfile=false` popularity, when I was looking into that, I couldn't find more information about it, besides that linked documentation.

None of the options that I explored were ideal, I think our case with an "optional" workspace is just uncommon, so the tools don't provide a good option to deal with that.
Inverting the submodule relationship between OSS and `e` probably would help (this seems to be a more popular scenario).

I'm fine with moving `e` deps to the root. At least, it won't be worse than what we have today and we can always revisit it in the future.

Thank you @ryanclark and @ravicious, your comments were really helpful.